pkg/server/test/write_authzmodel.go

```diff
@@ -428,6 +433,253 @@ func WriteAuthorizationModelTest(t *testing.T, datastore storage.OpenFGADatastor
 				fmt.Errorf("type 'user' defines a relation with an empty string for a name"),
 			),
 		},
+
+		// conditions
+		{
+			name: "condition_fails_undefined",
+			request: &openfgav1.WriteAuthorizationModelRequest{
+				StoreId: storeID,
+				TypeDefinitions: []*openfgav1.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "document",
+						Relations: map[string]*openfgav1.Userset{
+							"viewer": typesystem.This(),
+						},
+						Metadata: &openfgav1.Metadata{
+							Relations: map[string]*openfgav1.RelationMetadata{
+								"viewer": {
+									DirectlyRelatedUserTypes: []*openfgav1.RelationReference{
+										conditionedRelation(typesystem.WildcardRelationReference("user"), "invalid_condition_name"),
+									},
+								},
+							},
+						},
+					},
+				},
+				Conditions: map[string]*openfgav1.Condition{
+					"condition1": {
+						Name:       "condition1",
+						Expression: "param1 == 'ok'",
+						Parameters: map[string]*openfgav1.ConditionParamTypeRef{
+							"param1": {
+								TypeName: openfgav1.ConditionParamTypeRef_TYPE_NAME_STRING,
+							},
+						},
+					},
+				},
+			},
+			err: serverErrors.InvalidAuthorizationModelInput(
+				fmt.Errorf("condition invalid_condition_name is undefined for relation viewer"),
+			),
+		},
+		{
+			name: "condition_fails_syntax_error",
+			request: &openfgav1.WriteAuthorizationModelRequest{
+				StoreId: storeID,
+				TypeDefinitions: []*openfgav1.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "document",
+						Relations: map[string]*openfgav1.Userset{
+							"viewer": typesystem.This(),
+						},
+						Metadata: &openfgav1.Metadata{
+							Relations: map[string]*openfgav1.RelationMetadata{
+								"viewer": {
+									DirectlyRelatedUserTypes: []*openfgav1.RelationReference{
+										conditionedRelation(typesystem.WildcardRelationReference("user"), "condition1"),
+									},
+								},
+							},
+						},
+					},
+				},
+				Conditions: map[string]*openfgav1.Condition{
+					"condition1": {
+						Name:       "condition1",
+						Expression: "",
+						Parameters: map[string]*openfgav1.ConditionParamTypeRef{
+							"param1": {
+								TypeName: openfgav1.ConditionParamTypeRef_TYPE_NAME_STRING,
+							},
+						},
+					},
+				},
+			},
+			err: serverErrors.InvalidAuthorizationModelInput(
+				fmt.Errorf("failed to compile condition expression: ERROR: <input>:1:1: Syntax error: mismatched input '<EOF>' expecting {'[', '{', '(', '.', '-', '!', 'true', 'false', 'null', NUM_FLOAT, NUM_INT, NUM_UINT, STRING, BYTES, IDENTIFIER}"),
```

Comment:
We should sanitize the output here. This is a lot to consume IMO.

I may suggest we define some concrete errors in the `condition` package and return more succint errors here for the developer to consume.

e.g.

```go
if errors.Is(err, conditions.ErrCompilationFailed) {
    return serverErrors.InvalidAuthorizationModelnput(fmt.Errorf("failed to compile condition expression '%s'", expression))
}
```

If we want to be more specific I may suggest a different error model where we have the succint message and additional error details with the specific underlying cause "Syntax error: mismatched input '<EOF>' ..."

For example, maybe we consider using the standard gRPC error design:
https://grpc.io/docs/guides/error/#richer-error-model
https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto

Reply:
took a simpler stab at this for now so we can continue moving forward.

> For example, maybe we consider using something more like the standard gRPC error design (or use it directly):
https://grpc.io/docs/guides/error/#richer-error-model
https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto

💯